contrib/lustre_plugin.c

```diff
@@ -0,0 +1,262 @@
+        uint16_t stripe_count) {
+    for (int i = 0; i < stripe_count; i++) {
+        uint32_t ost_index = stripe_array[i].l_ost_idx;
+        if (grow_stripe_tracker(s, ost_index)) {
+            /* Just give up if we couldn't grow the stripe array large enough :( */
+            fprintf(stderr, "lustre plugin: could not allocate memory, information may be incomplete");
+            return;
+        }
+
+        s->stripe_count[ost_index] += 1;
+    }
+}
+
+/*
+ * Set up initial state for tracking Lustre stripe info.
+ */
+void *db_init(sqlite3 *db) {
+    struct stripe_tracker *state = new_stripe_tracker();
+    if (!state) {
+        fprintf(stderr, "lustre plugin: could not allocate memory to track stripe info");
+        return NULL;
+    }
+
+    sqlite3_stmt *statement;
+    char *text = "CREATE TABLE lustre_summary (ost_index INTEGER PRIMARY KEY, num_files INTEGER);";
+
+    int res = sqlite3_prepare_v2(db, text, strlen(text) + 1, &statement, NULL);
+    if (res != SQLITE_OK) {
+        fprintf(stderr, "lustre plugin: error preparing sqlite3 statement: %d\n", res);
+        return NULL;
+    }
+
+    res = sqlite3_step(statement);
+    if (res != SQLITE_DONE) {
+        fprintf(stderr, "lustre plugin: error evaluting statement: %d\n", res);
+        return NULL;
+    }
+
+    sqlite3_finalize(statement);
+
+    return state;
+}
+
+/*
+ * Save stripe tracking info to the database and clean up state.
+ */
+void db_exit(sqlite3 *db, void *user_data) {
+    struct stripe_tracker *state = (struct stripe_tracker *) user_data;
+
+    for (uint32_t i = 0; i <= state->num_osts; i++) {
+        sqlite3_stmt *statement;
+        char *text = sqlite3_mprintf("INSERT INTO %s VALUES(%" PRIu32 ", %" PRIu64 ");",
+                "lustre_summary", i, state->stripe_count[i]);
+        int res = sqlite3_prepare_v2(db, text, -1, &statement, NULL);
+        if (res != SQLITE_OK) {
+            fprintf(stderr, "lustre plugin: error preparing sqlite3 statement: %d\n", res);
+            goto out;
+        }
+
+        res = sqlite3_step(statement);
+        if (res != SQLITE_DONE) {
+            fprintf(stderr, "lustre plugin: error evaluting statement: %d\n", res);
+        }
+
+        sqlite3_finalize(statement);
+        sqlite3_free(text);
+    }
+
+  out:
+    destroy_stripe_tracker(state);
+};
+
+static void *alloc_lum() {
+    size_t v1 = sizeof(struct lov_user_md_v1) +
+        LOV_MAX_STRIPE_COUNT * sizeof(struct lov_user_ost_data_v1);
+    size_t v3 = sizeof(struct lov_user_md_v3) +
+        LOV_MAX_STRIPE_COUNT * sizeof(struct lov_user_ost_data_v1);
+
+    return calloc(1, v1 > v3 ? v1 : v3);
+}
+
+void process_file(char *path, sqlite3 *db, void *user_data) {
+    sqlite3_stmt *statement;
+
+    struct lov_user_md *layout_info = alloc_lum();
+
+    int rc = llapi_file_get_stripe(path, layout_info);
+
+    if (rc) {
+        printf("lustre plugin: error getting stripe info for %s: %s\n", path, strerror(errno));
```

Comment:
`stderr`

Reply:
`errno` should be copied before being used

Reply:
While in this case you're fine, people tend to call other functions that they don't realize change `errno`, and end up getting prints that say that there are no errors even though they entered an error block. The way to prevent this is to copy `errno` so that side-effects don't mess up your error handling. I forget what I was doing, but I've seen this happen recently.